File: tideform-swift/Sources/Tideform/Lib/Walrus.swift
//
//  Walrus.swift
//  Tideform · Lib layer
//
//  Walrus blob I/O, mirroring the Expo `lib/walrus.ts` surface:
//    readBlob(id), readJson(id), uploadBlob(bytes,{owner}), uploadJson(obj,{owner}), blobUrl(id)
//
//  READS  — public aggregator, no auth, straight from the device (source-of-truth §5).
//  WRITES — sponsored via the Zentos backend `POST /api/walrus/upload` (multipart);
//           the user pays zero WAL/SUI and the sponsor API key never reaches the client.
//
//  IMPORTANT: the returned `blobId` is an ASCII string. It is stored on-chain as
//  `vector<u8>` of its ASCII bytes (Move.swift encodes it that way) — never base64.
//

import Foundation

public enum WalrusError: Error, CustomStringConvertible {
    case http(status: Int, body: String)
    case badResponse(String)

    public var description: String {
        switch self {
        case .http(let s, let b): return "Walrus HTTP \(s): \(b)"
        case .badResponse(let m): return "Walrus bad response: \(m)"
        }
    }
}

/// Result of a sponsored upload (`POST /api/walrus/upload`). `blobId` is what you
/// store on-chain.
public struct WalrusUploadResult: Decodable, Sendable {
    public let blobId: String
    public let sponsoredBlobId: String?
    public let txDigest: String?
    public let endEpoch: Int?
    public let walCost: JSONValue?

    private enum CodingKeys: String, CodingKey {
        case blobId = "blob_id"
        case sponsoredBlobId = "sponsored_blob_id"
        case txDigest = "tx_digest"
        case endEpoch = "end_epoch"
        case walCost = "wal_cost"
    }
}

public final class Walrus: @unchecked Sendable {

    public static let shared = Walrus()

    private let aggregator: String
    private let backendBaseUrl: String
    /// Cookie-aware session so the sponsored upload carries the Zentos session cookie.
    private let session: URLSession

    public init(
        aggregator: String = env.walrusAggregator,
        backendBaseUrl: String = env.backendBaseUrl,
        session: URLSession = tideformURLSession
    ) {
        self.aggregator = aggregator.hasSuffix("/") ? String(aggregator.dropLast()) : aggregator
        self.backendBaseUrl =
            backendBaseUrl.hasSuffix("/") ? String(backendBaseUrl.dropLast()) : backendBaseUrl
        self.session = session
    }

    // MARK: - Reads (public aggregator)

    /// Public read URL for a blob — useful for AsyncImage / direct linking.
    public func blobUrl(_ id: String) -> URL {
        URL(string: "\(aggregator)/v1/blobs/\(id)")!
    }

    /// Raw blob bytes.
    public func readBlob(_ id: String) async throws -> Data {
        let (data, response) = try await session.data(from: blobUrl(id))
        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw WalrusError.http(
                status: http.statusCode, body: String(data: data, encoding: .utf8) ?? "")
        }
        return data
    }

    /// Decode a blob as JSON into a concrete `Decodable` (e.g. `FormSchema`, `Submission`).
    public func readJson<T: Decodable>(_ type: T.Type = T.self, id: String) async throws -> T {
        let data = try await readBlob(id)
        return try JSONDecoder().decode(T.self, from: data)
    }

    /// Decode a blob as untyped JSON.
    public func readJson(id: String) async throws -> JSONValue {
        let data = try await readBlob(id)
        return try JSONDecoder().decode(JSONValue.self, from: data)
    }

    // MARK: - Writes (sponsored, via backend)

    /// Upload raw bytes through the sponsored backend route. Returns the on-chain `blobId`.
    /// `owner` is the user's Sui address (`creator_address`).
    @discardableResult
    public func uploadBlob(
        _ bytes: Data,
        owner: String,
        epochs: Int = 5,
        deletable: Bool = true,
        filename: String = "blob.bin",
        contentType: String = "application/octet-stream"
    ) async throws -> WalrusUploadResult {
        let url = URL(string: "\(backendBaseUrl)/api/walrus/upload")!
        let boundary = "Boundary-\(UUID().uuidString)"

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue(
            "multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        req.httpBody = Walrus.multipartBody(
            boundary: boundary,
            fields: [
                "creator_address": owner,
                "epochs": String(epochs),
                "deletable": deletable ? "true" : "false",
            ],
            fileField: "file",
            filename: filename,
            contentType: contentType,
            fileData: bytes)

        let (data, response) = try await session.data(for: req)
        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw WalrusError.http(
                status: http.statusCode, body: String(data: data, encoding: .utf8) ?? "")
        }
        do {
            return try JSONDecoder().decode(WalrusUploadResult.self, from: data)
        } catch {
            throw WalrusError.badResponse("\(error)")
        }
    }

    /// Encode an `Encodable` (e.g. `FormSchema`, `Submission`) to JSON and upload it.
    @discardableResult
    public func uploadJson<T: Encodable>(
        _ object: T,
        owner: String,
        epochs: Int = 5,
        deletable: Bool = true
    ) async throws -> WalrusUploadResult {
        let data = try JSONEncoder().encode(object)
        return try await uploadBlob(
            data, owner: owner, epochs: epochs, deletable: deletable,
            filename: "payload.json", contentType: "application/json")
    }

    // MARK: - Multipart encoding

    private static func multipartBody(
        boundary: String,
        fields: [String: String],
        fileField: String,
        filename: String,
        contentType: String,
        fileData: Data
    ) -> Data {
        var body = Data()
        let crlf = "\r\n"
        func append(_ s: String) { body.append(s.data(using: .utf8)!) }

        for (name, value) in fields {
            append("--\(boundary)\(crlf)")
            append("Content-Disposition: form-data; name=\"\(name)\"\(crlf)\(crlf)")
            append("\(value)\(crlf)")
        }

        append("--\(boundary)\(crlf)")
        append(
            "Content-Disposition: form-data; name=\"\(fileField)\"; filename=\"\(filename)\"\(crlf)")
        append("Content-Type: \(contentType)\(crlf)\(crlf)")
        body.append(fileData)
        append(crlf)
        append("--\(boundary)--\(crlf)")
        return body
    }
}

/// Module-level alias mirroring the JS `walrus` singleton (`walrus.readJson(...)`).
public let walrus = Walrus.shared

File: tideform-expo/lib/walrus.ts
/**
 * lib/walrus.ts — Walrus blob I/O (source-of-truth §5).
 *
 *  READS  → public aggregator, no auth, straight from the device.
 *             GET ${WALRUS_AGGREGATOR}/v1/blobs/{blobId} → raw bytes
 *
 *  WRITES → SPONSORED via the Zentos backend (zero WAL/SUI for the user).
 *             POST multipart/form-data ${BACKEND}/api/walrus/upload
 *             fields: file, creator_address, epochs=5, deletable=true
 *             → { blob_id, sponsored_blob_id, tx_digest, end_epoch?, wal_cost? }
 *           The returned `blob_id` is what you store ON-CHAIN.
 *
 * This is the `walrus` member of the shared lib contract:
 *   readBlob, readJson, uploadBlob, uploadJson, blobUrl
 */

import { toBase64 } from '@mysten/sui/utils';
import * as FileSystem from 'expo-file-system';

import { cookieFetch } from './cookies';
import { env } from './env';

/** Public aggregator URL for a blob (use directly as an <Image> source, etc.). */
export function blobUrl(blobId: string): string {
  return `${env.walrusAggregator}/v1/blobs/${blobId}`;
}

/** Read a blob's raw bytes from the public aggregator. */
export async function readBlob(blobId: string): Promise<Uint8Array> {
  const res = await fetch(blobUrl(blobId));
  if (!res.ok) {
    throw new Error(`Walrus read failed for ${blobId}: ${res.status} ${res.statusText}`);
  }
  const buf = await res.arrayBuffer();
  return new Uint8Array(buf);
}

/** Read a blob and parse it as JSON. */
export async function readJson<T = unknown>(blobId: string): Promise<T> {
  const res = await fetch(blobUrl(blobId));
  if (!res.ok) {
    throw new Error(`Walrus read failed for ${blobId}: ${res.status} ${res.statusText}`);
  }
  return (await res.json()) as T;
}

export interface UploadOptions {
  /** Sui address recorded as the blob's creator (the signed-in user). */
  owner: string;
  /** Storage duration in Walrus epochs. Backend default mirrors the web app (5). */
  epochs?: number;
  /** Whether the blob is deletable. Defaults to true (matches the web app). */
  deletable?: boolean;
  /** MIME type for the multipart part (informational). */
  mime?: string;
  /** File name for the multipart part. */
  name?: string;
}

export interface UploadResult {
  /** The canonical Walrus blob ID — STORE THIS ON-CHAIN. */
  blobId: string;
  /** The sponsor-side blob ID (sponsored copy), if returned. */
  sponsoredBlobId?: string;
  /** Sponsor registration/certify tx digest. */
  txDigest?: string;
  endEpoch?: number;
  walCost?: number;
}

/** Backend JSON shape (snake_case) → normalized camelCase UploadResult. */
function normalizeUpload(json: Record<string, unknown>): UploadResult {
  return {
    blobId: String(json.blob_id ?? ''),
    sponsoredBlobId:
      json.sponsored_blob_id != null ? String(json.sponsored_blob_id) : undefined,
    txDigest: json.tx_digest != null ? String(json.tx_digest) : undefined,
    endEpoch: json.end_epoch != null ? Number(json.end_epoch) : undefined,
    walCost: json.wal_cost != null ? Number(json.wal_cost) : undefined,
  };
}

/**
 * Upload raw bytes via the sponsored backend route. The sponsor key lives only
 * on the server; the user pays nothing.
 *
 * RN gotcha (source-of-truth §12 / expo-sui skill): React Native's `FormData`
 * cannot reliably stream raw bytes or a `Blob` — the multipart body comes through
 * empty. The proven path is to stage the bytes to a cache file with
 * `expo-file-system`, then append a `{ uri, name, type }` file part. We write the
 * bytes as base64 (lossless for binary) and clean the temp file up afterwards.
 */
export async function uploadBlob(
  bytes: Uint8Array,
  opts: UploadOptions,
): Promise<UploadResult> {
  const name = opts.name ?? 'submission.json';
  const mime = opts.mime ?? 'application/octet-stream';

  // Stage the bytes to a cache file so RN can attach a real file part.
  const fileUri = `${FileSystem.cacheDirectory}tideform-upload-${Date.now()}`;
  await FileSystem.writeAsStringAsync(fileUri, toBase64(bytes), {
    encoding: FileSystem.EncodingType.Base64,
  });

  try {
    const form = new FormData();
    // `as any` — RN's file-part shape isn't in the DOM FormData typings.
    form.append('file', { uri: fileUri, name, type: mime } as any);
    form.append('creator_address', opts.owner);
    form.append('epochs', String(opts.epochs ?? 5));
    form.append('deletable', String(opts.deletable ?? true));

    // Do NOT set Content-Type — the platform sets the multipart boundary itself.
    const res = await cookieFetch(`${env.backendBaseUrl}/api/walrus/upload`, {
      method: 'POST',
      body: form,
    });
    if (!res.ok) {
      const detail = await res.text().catch(() => '');
      throw new Error(`Walrus upload failed: ${res.status} ${res.statusText} ${detail}`);
    }
    return normalizeUpload((await res.json()) as Record<string, unknown>);
  } finally {
    // Best-effort cleanup of the staged temp file.
    void FileSystem.deleteAsync(fileUri, { idempotent: true }).catch(() => {});
  }
}

/** Serialize an object to JSON bytes and upload it. */
export async function uploadJson(
  obj: unknown,
  opts: UploadOptions,
): Promise<UploadResult> {
  const bytes = new TextEncoder().encode(JSON.stringify(obj));
  return uploadBlob(bytes, {
    mime: 'application/json',
    name: 'payload.json',
    ...opts,
  });
}
